notebooks/Ensemble parameter exploration.ipynb

```diff
@@ -0,0 +1,1124 @@
+       "    </tr>\n",
+       "  </tbody>\n",
+       "</table>\n",
+       "</div>"
+      ],
+      "text/plain": [
+       "                Type                                         model_name  \\\n",
+       "411   Ensemble model  (bert_log_reg_bert_210221, bert_SVM_bert_21022...   \n",
+       "859   Ensemble model  (bert_naive_bayes_bert_210221, bert_log_reg_be...   \n",
+       "1831  Ensemble model  (bert_log_reg_bert_210221, bert_SVM_bert_21022...   \n",
+       "3098  Ensemble model  (bert_naive_bayes_bert_210221, bert_log_reg_be...   \n",
+       "\n",
+       "      num_models  num_agree        f1  precision_score  recall_score  \n",
+       "411            3          2  0.830189         0.846154      0.814815  \n",
+       "859            4          3  0.830189         0.846154      0.814815  \n",
+       "1831           4          3  0.830189         0.846154      0.814815  \n",
+       "3098           5          4  0.830189         0.846154      0.814815  "
+      ]
+     },
+     "execution_count": 21,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "best_results = all_combos_scores_df.loc[\n",
+    "    (\n",
+    "        (all_combos_scores_df['precision_score'] >= precision_threshold) & \n",
+    "        (all_combos_scores_df['recall_score'] >= recall_threshold)\n",
+    "    )\n",
+    "]\n",
+    "best_results"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 24,
+   "id": "bacterial-court",
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "[('bert_log_reg_bert_210221',\n",
+       "  'bert_SVM_bert_210221',\n",
+       "  'count_naive_bayes_210221'),\n",
+       " ('bert_naive_bayes_bert_210221',\n",
+       "  'bert_log_reg_bert_210221',\n",
+       "  'bert_SVM_bert_210221',\n",
+       "  'count_naive_bayes_210221'),\n",
+       " ('bert_log_reg_bert_210221',\n",
+       "  'bert_SVM_bert_210221',\n",
+       "  'bert_naive_bayes_scibert_210221',\n",
+       "  'count_naive_bayes_210221'),\n",
+       " ('bert_naive_bayes_bert_210221',\n",
+       "  'bert_log_reg_bert_210221',\n",
+       "  'bert_SVM_bert_210221',\n",
+       "  'bert_naive_bayes_scibert_210221',\n",
+       "  'count_naive_bayes_210221')]"
+      ]
+     },
+     "execution_count": 24,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "best_results['model_name'].tolist()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 22,
+   "id": "assigned-heavy",
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "['bert_log_reg_bert_210221',\n",
+       " 'bert_SVM_bert_210221',\n",
+       " 'count_naive_bayes_210221']"
+      ]
+     },
+     "execution_count": 22,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# Less models to have to make predictions from the better\n",
+    "optimal_model_i = 411\n",
```

Comment:
I don't get it where 411 comes from?

Reply:
oh yes, I'll add my logic here instead of hardcoding. I've just manually picked which my favourite combination from the dataframe (row 411). But that was because it has the fewest number of models, so I can add this logic in instead